De-abuse TyKind::Error in exhaustiveness checking

Nadrieril opened the pull request:
Replaces https://github.com/rust-lang/rust/pull/71074. Context: https://github.com/rust-lang/rust/issues/70866.

In order to remove the use of `TyKind::Error`, I had to make sure we skip over those fields whose inhabitedness should not be observed. This is potentially error-prone however, since we must be careful not to mix filtered and unfiltered lists of patterns. I managed to hide away most of the filtering behind a new `Fields` struct, that I used everywhere relevant. I quite like the result; I think the twin concepts of `Constructor` and `Fields` make a good mental model.

As usual, I tried to separate commits that shuffle code around from commits that require more thought to review.

cc @varkor @Centril

Nadrieril committed:
Factor the code that generates TyErrs

Nadrieril committed:
We already handle arrays of unknown length correctly

Nadrieril committed:
Only need TyErr for uninhabited types

rust-highfive commented:
r? @matthewjasper

(rust_highfive has picked a reviewer for you, use r? to override)

Nadrieril commented:
May I request a perf run on this? I'm not confident I managed to preserve performance

jonas-schievink commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit e5667234a7fe2ca921766a197933229586749832 with merge 10bf6df2a72191317fadc1859f3d849707705018...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e5667234a7fe2ca921766a197933229586749832","merge_sha":"10bf6df2a72191317fadc1859f3d849707705018"} -->

rust-highfive commented:
The job `x86_64-gnu-llvm-8` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=28439&view=logs&jobId=698b0e4a-0480-533d-81fa-f38599aa328e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/28439/logs/71), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/28439/logs/71)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/20ec7cb6-41fa-4b62-ad2f-fb8a23ab3d2c.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/71930/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/71930/merge:refs/remotes/pull/71930/merge
---
 ---> cb2676f08729
Step 5/8 : ENV RUST_CONFIGURE_ARGS       --build=x86_64-unknown-linux-gnu       --llvm-root=/usr/lib/llvm-8       --enable-llvm-link-shared       --set rust.thin-lto-import-instr-limit=10
 ---> Using cache
 ---> df25ce111862
Step 6/8 : ENV SCRIPT python2.7 ../x.py test --exclude src/tools/tidy &&            python2.7 ../x.py test src/test/mir-opt --pass=build                                   --target=armv5te-unknown-linux-gnueabi &&            python2.7 ../x.py test src/tools/tidy
 ---> 599b9ac96b27
Step 7/8 : ENV NO_DEBUG_ASSERTIONS=1
 ---> Using cache
 ---> 091087e35a36
---
   Compiling rustc_feature v0.0.0 (/checkout/src/librustc_feature)
   Compiling fmt_macros v0.0.0 (/checkout/src/libfmt_macros)
   Compiling rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
   Compiling rustc_hir v0.0.0 (/checkout/src/librustc_hir)
   Compiling rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
   Compiling rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
   Compiling rustc_parse v0.0.0 (/checkout/src/librustc_parse)
   Compiling rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
   Compiling rustc_ast_passes v0.0.0 (/checkout/src/librustc_ast_passes)
---
   Compiling rustc_feature v0.0.0 (/checkout/src/librustc_feature)
   Compiling fmt_macros v0.0.0 (/checkout/src/libfmt_macros)
   Compiling rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
   Compiling rustc_hir v0.0.0 (/checkout/src/librustc_hir)
   Compiling rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
   Compiling rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
   Compiling rustc_parse v0.0.0 (/checkout/src/librustc_parse)
   Compiling rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
   Compiling rustc_ast_passes v0.0.0 (/checkout/src/librustc_ast_passes)
---
......................i............................................................................. 1800/9988
.................................................................................................... 1900/9988
.......................................i............................................................ 2000/9988
.................................................................................................... 2100/9988
.............................iiiii.................................................................. 2200/9988
.................................................................................................... 2400/9988
.................................................................................................... 2500/9988
.................................................................................................... 2600/9988
.................................................................................................... 2700/9988
---
................i...............i................................................................... 5100/9988
.................................................................................................... 5200/9988
..............................................................i..................................... 5300/9988
.....................................................i.............................................. 5400/9988
.........................................................ii.ii........i...i......................... 5500/9988
..................................................................................................i. 5600/9988
....i............................................................................................... 5800/9988
........................................ii.....................................i.................... 5900/9988
.................................................................................................... 6000/9988
.................................................................................................... 6100/9988
.................................................................................................... 6100/9988
............................................................................ii...i..ii...........i.. 6200/9988
.................................................................................................... 6400/9988
.................................................................................................... 6500/9988
.................................................................................................... 6600/9988
.................................................................................................... 6600/9988
........i..ii....................................................................................... 6700/9988
.................................................................................................... 6900/9988
.........i.......................................................................................... 7000/9988
.................................................................................................... 7100/9988
...................................................i................................................ 7200/9988
---
.................................................................................................... 7900/9988
.................................................................................................... 8000/9988
.................................................................................................... 8100/9988
...................i................................................................................ 8200/9988
.........................................................................iiiiii.iiiii..i............ 8300/9988
..........................i......................................................................... 8500/9988
.................................................................................................... 8600/9988
.................................................................................................... 8700/9988
.................................................................................................... 8800/9988
---
---- [run-fail] run-fail/match-disc-bot.rs stdout ----

error: compilation failed!
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-fail/match-disc-bot.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-fail/match-disc-bot/a" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-fail/match-disc-bot/auxiliary"
------------------------------------------

------------------------------------------
stderr:
---
thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:348:22
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/run-fail" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-fail" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "run-fail" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-8/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "8.0.0" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --exclude src/tools/tidy
Build completed unsuccessfully in 1:02:08
Build completed unsuccessfully in 1:02:08
== clock drift check ==
  local time: Tue May  5 22:04:42 UTC 2020
  network time: Tue, 05 May 2020 22:04:42 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/71930/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/71930/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (4211) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/647686941), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=28443)
Build commit: 10bf6df2a72191317fadc1859f3d849707705018 (`10bf6df2a72191317fadc1859f3d849707705018`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=28443","checks-actions":"https://github.com/rust-lang-ci/rust/runs/647686941"},"merge_sha":"10bf6df2a72191317fadc1859f3d849707705018"} -->

rust-timer commented:
Queued 10bf6df2a72191317fadc1859f3d849707705018 with parent f8d394e5184fe3af761ea1e5ba73f993cfb36dfe, future [comparison URL](https://perf.rust-lang.org/compare.html?start=f8d394e5184fe3af761ea1e5ba73f993cfb36dfe&end=10bf6df2a72191317fadc1859f3d849707705018).

mark-i-m commented:
cc @eddyb  (who I was working with on the tykind::error effort)

mark-i-m commented:
It looks like the CI failure is a bug:

```
warning: unreachable arm
 --> /checkout/src/test/run-fail/match-disc-bot.rs:7:17
  |
6 |     match f() {
  |           --- any code following this expression is unreachable
7 |         true => 1,
  |                 ^ unreachable arm
  |
  = note: `#[warn(unreachable_code)]` on by default

error[E0004]: non-exhaustive patterns: `_` not covered
 --> /checkout/src/test/run-fail/match-disc-bot.rs:6:11
  |
6 |     match f() {
  |           ^^^ pattern `_` not covered
  |
  = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
  = note: the matched value is of type `!`

error: aborting due to previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0004`.
```

mark-i-m commented on src/librustc_mir_build/hair/pattern/_match.rs:
```diff
@@ -861,107 +860,10 @@ impl<'tcx> Constructor<'tcx> {
     /// must have as many elements as this constructor's arity.
     ///
-    /// This must be consistent with `wildcard_subpatterns`, `specialize_one_pattern`, and `arity`.
+    /// This is morally the inverse of `specialize_one_pattern`.
```
```suggestion
    /// This is conceptually the inverse of `specialize_one_pattern`.
```

This use of "morally" tends to be a bit confusing to non-English speakers (and even as a native English speaker, I tend to find it only vaguely defined)...

Nadrieril replied:
How about "roughly"?

mark-i-m commented on src/librustc_mir_build/hair/pattern/_match.rs:
```diff
@@ -2344,39 +2381,44 @@ fn patterns_for_variant<'p, 'tcx>(
 /// Structure patterns with a partial wild pattern (Foo { a: 42, .. }) have their missing
 /// fields filled with wild patterns.
+///
+/// This is morally the inverse of `Constructor::apply`.
```
```suggestion
/// This is conceptually the inverse of `Constructor::apply`.
```

rust-timer commented:
Finished benchmarking try commit 10bf6df2a72191317fadc1859f3d849707705018, [comparison URL](https://perf.rust-lang.org/compare.html?start=f8d394e5184fe3af761ea1e5ba73f993cfb36dfe&end=10bf6df2a72191317fadc1859f3d849707705018).

mark-i-m commented:
Perf looks neutral

Nadrieril commented:
The regression was caused by https://github.com/rust-lang/rust/pull/71930/commits/624ff54529df551e13e299d4369d3418c9db0f9b . The regressed test is as follows:
```rust
fn f() -> ! {
    panic!("quux")
}
fn g() -> isize {
    match f() {
        true => 1,
        false => 0,
    }
}
```
The reason it now fails is because instead of using the type of the first pattern to do our computations, we now use the type of the scrutinee. But that shouldn't make a difference, should it? Somehow here the scrutinee is given the type `!`; I would expect type inference to give it the type `bool` though?
In fact, the following variation works. In that one, `x` is correctly given the type `bool`, so exhaustiveness works as expected.
```rust
fn g() -> isize {
	let x = f();
    match x {
        true => 1,
        false => 0,
    }
}
```
So I don't think it's my fault; that looks like a type inference issue to me. For context, the type of the scrutinee is retrieved here https://github.com/rust-lang/rust/blob/8da5869fb738e51438dd1e0697c1b2c84eb11c59/src/librustc_mir_build/hair/pattern/check_match.rs#L181

Nadrieril committed:
Fix incorrect ordering

I introduced this mistake in 175976e2a2b03c3f347d4eff28661445c3c58372
and I can't quite remember what the reasoning was back then. I think I
modeled it on `apply_constructor`, not realizing there was an important
difference in the order in which fields were stored.

matthewjasper commented:
`node_type` should be `expr_ty_adjusted(scrut)`

Nadrieril commented:
@matthewjasper Thanks, that was indeed the issue!

rust-highfive commented:
The job `x86_64-gnu-llvm-8` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=28508&view=logs&jobId=698b0e4a-0480-533d-81fa-f38599aa328e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/28508/logs/71), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/28508/logs/71)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/98ee66e6-9fc3-4f8d-bedc-b090433b2610.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/71930/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/71930/merge:refs/remotes/pull/71930/merge
---
 ---> cb2676f08729
Step 5/8 : ENV RUST_CONFIGURE_ARGS       --build=x86_64-unknown-linux-gnu       --llvm-root=/usr/lib/llvm-8       --enable-llvm-link-shared       --set rust.thin-lto-import-instr-limit=10
 ---> Using cache
 ---> df25ce111862
Step 6/8 : ENV SCRIPT python2.7 ../x.py test --exclude src/tools/tidy &&            python2.7 ../x.py test src/test/mir-opt --pass=build                                   --target=armv5te-unknown-linux-gnueabi &&            python2.7 ../x.py test src/tools/tidy
 ---> 599b9ac96b27
Step 7/8 : ENV NO_DEBUG_ASSERTIONS=1
 ---> Using cache
 ---> 091087e35a36
---
   Compiling rustc_feature v0.0.0 (/checkout/src/librustc_feature)
   Compiling fmt_macros v0.0.0 (/checkout/src/libfmt_macros)
   Compiling rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
   Compiling rustc_hir v0.0.0 (/checkout/src/librustc_hir)
   Compiling rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
   Compiling rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
   Compiling rustc_parse v0.0.0 (/checkout/src/librustc_parse)
   Compiling rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
   Compiling rustc_ast_passes v0.0.0 (/checkout/src/librustc_ast_passes)
---
  local time: Wed May  6 13:52:51 UTC 2020
  network time: Wed, 06 May 2020 13:52:51 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/71930/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/71930/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (3611) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Nadrieril commented:
@matthewjasper Actually, your suggestion regressed this:
```rust
fn from(never: !) -> TryFromIntError {
    match never {}
}
```
Apparently `expr_ty_adjusted` causes `never` to get type `()` here :confused:

mark-i-m commented:
Indeed, we confirmed that this works without 624ff54 in #71939 ... Would it be reasonable to merge this PR without that commit and fix the error in a followup?

rust-highfive commented:
The job `x86_64-gnu-llvm-8` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=28621&view=logs&jobId=698b0e4a-0480-533d-81fa-f38599aa328e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/28621/logs/71), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/28621/logs/71)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/80a9cd0f-5ee0-429a-9c31-af7ac2c7e311.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/71930/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/71930/merge:refs/remotes/pull/71930/merge
---
 ---> cb2676f08729
Step 5/8 : ENV RUST_CONFIGURE_ARGS       --build=x86_64-unknown-linux-gnu       --llvm-root=/usr/lib/llvm-8       --enable-llvm-link-shared       --set rust.thin-lto-import-instr-limit=10
 ---> Using cache
 ---> df25ce111862
Step 6/8 : ENV SCRIPT python2.7 ../x.py test --exclude src/tools/tidy &&            python2.7 ../x.py test src/test/mir-opt --pass=build                                   --target=armv5te-unknown-linux-gnueabi &&            python2.7 ../x.py test src/tools/tidy
 ---> 599b9ac96b27
Step 7/8 : ENV NO_DEBUG_ASSERTIONS=1
 ---> Using cache
 ---> 091087e35a36
---
   Compiling rustc_feature v0.0.0 (/checkout/src/librustc_feature)
   Compiling fmt_macros v0.0.0 (/checkout/src/libfmt_macros)
   Compiling rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
   Compiling rustc_hir v0.0.0 (/checkout/src/librustc_hir)
   Compiling rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
   Compiling rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
   Compiling rustc_parse v0.0.0 (/checkout/src/librustc_parse)
   Compiling rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
   Compiling rustc_ast_passes v0.0.0 (/checkout/src/librustc_ast_passes)
---
  local time: Thu May  7 16:38:47 UTC 2020
  network time: Thu, 07 May 2020 16:38:47 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/71930/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/71930/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (3960) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

mark-i-m commented:
```
 error[E0004]: non-exhaustive patterns: type `()` is non-empty
    --> src/libcore/num/mod.rs:5046:15
     |
5046 |         match never {}
     |               ^^^^^
     |
     = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
     = note: the matched value is of type `()`

error: aborting due to previous error
```

Nadrieril commented:
Yeah I didn't mean to push that yet, I'm still hoping that @matthewjasper might pop in and figure out the arcane invocation needed to get the correct type x)

matthewjasper commented:
The problem is that the call here is causing the type of the scrutinee to become a diverging variable, and fall back to `()`. https://github.com/rust-lang/rust/blob/7b805396bf46dce972692a6846ce2ad8481c5f85/src/librustc_typeck/check/_match.rs#L439-L447
Maybe adding this right above the linked code will result in the adjusted type being correct.
````
else if arms.is_empty() {
    self.check_expr(scrut)
```

bors commented:
:umbrella: The latest upstream changes (presumably #72021) made this pull request unmergeable. Please resolve the merge conflicts.

Nadrieril committed:
Factor out a struct that holds subfields of a pattern

Nadrieril committed:
Use Fields as output to specialize_one_pattern

Nadrieril committed:
Clarify specialize_one_pattern

Using a more error-oriented approache to `Option`.

Nadrieril committed:
Make all field-handling go through Fields

Nadrieril committed:
Filter out fields that should not be seen

This was previously done by giving those fields the type tyerr. This was
a hack.

Nadrieril committed:
We don't use tyerr anymore

This however unearthed a bug, hence the FIXME and the workaround.

Nadrieril committed:
Small allocation improvement

Nadrieril committed:
Fix bug just discovered

Suggested by matthewjasper here:
https://github.com/rust-lang/rust/pull/71930#issuecomment-626022502
I have no idea what this does but it seems to work.

Nadrieril committed:
typo

Nadrieril commented:
@matthewjasper Thanks a lot, that indeed fixed it!

I've cleaned up the commits noticeably, and hopefully improved perf a bit. That should be ready to be reviewed then.

matthewjasper commented:
I don't really understand the exhaustiveness checking code. Would @varkor be able to review this?

varkor commented:
r? @varkor

varkor commented on src/librustc_mir_build/hair/pattern/_match.rs:
```diff
@@ -1033,8 +932,246 @@ impl<'tcx> Constructor<'tcx> {
+        pats.into_iter()
+    }
+
+    /// Overrides some of the fields with the provided patterns.
```
It'd be nice to expand this comment a little (and on `replace_fields_indexed` / just point the comment on `replace_fields_indexed` to this one) to explain in what situation it is used (so someone doesn't have to scout around for the context).

varkor reviewed (commented):
This approach seems reasonable to me, though I think I want to spend a bit more time looking over it. The comment on `Fields` is helpful. It might also be worth adding some info on this change to the large explanation comment at the top of the file?

It always takes me some time to reacquaint myself with this file, because the algorithm has grown quite a lot since its creation. I'm beginning to sway towards desiring more detailed comments on each function, just to give a reader some idea of the context if they take a look at an arbitrary function, without having to work through from top to bottom. Doing this exhaustively is tangential to this pull request (and overall, the file is fairly well-commented as it is), but I've highlighted a couple of functions that I think could do with being expanded upon a little more, which are relevant to this pull request.

It's always worth double-checking that there aren't performance regressions when it comes to the pattern matching code, so I'm going to try a perf run once more, just to make sure.

varkor commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 19bfef8af4eb0c618d0bd03c0034db00d65cd1f3 with merge 2be296f7e04b37e0e7fb4218bd752868b299f795...
<!-- homu: {"type":"TryBuildStarted","head_sha":"19bfef8af4eb0c618d0bd03c0034db00d65cd1f3","merge_sha":"2be296f7e04b37e0e7fb4218bd752868b299f795"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=29308)
Build commit: 2be296f7e04b37e0e7fb4218bd752868b299f795 (`2be296f7e04b37e0e7fb4218bd752868b299f795`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=29308"},"merge_sha":"2be296f7e04b37e0e7fb4218bd752868b299f795"} -->

rust-timer commented:
Queued 2be296f7e04b37e0e7fb4218bd752868b299f795 with parent ed084b0b8341c974769a0328f61851b0e1fc17fa, future [comparison URL](https://perf.rust-lang.org/compare.html?start=ed084b0b8341c974769a0328f61851b0e1fc17fa&end=2be296f7e04b37e0e7fb4218bd752868b299f795).

rust-timer commented:
Finished benchmarking try commit 2be296f7e04b37e0e7fb4218bd752868b299f795, [comparison URL](https://perf.rust-lang.org/compare.html?start=ed084b0b8341c974769a0328f61851b0e1fc17fa&end=2be296f7e04b37e0e7fb4218bd752868b299f795).

Nadrieril committed:
Improve comments

Nadrieril committed:
Cache len in Fields

Nadrieril commented:
@varkor I couldn't reproduce the perf impact locally, so I rebased on current master but I still don't see it. On my machine I see a 1% _improvement_ on the unicode_normalization tests :/
I improved the comments as you suggested.

varkor commented on src/librustc_mir_build/hair/pattern/_match.rs:
```diff
@@ -1033,8 +978,263 @@ impl<'tcx> Constructor<'tcx> {
+    }
+}
+
+/// Some fields need to be explicitely hidden away in certain cases; see the comment above the
```
```suggestion
/// Some fields need to be explicitly hidden away in certain cases; see the comment above the
```

varkor reviewed (commented):
Thanks @Nadrieril, these comments are great :) I spotted one typo, but apart from that, I think this is ready to go.

varkor commented:
I'm inclined to treat the perf results as negligible, especially if you see an improvement lately, but we'll make sure the PR isn't rolled up, so if it does have an effect, it'll be easier to track down.

@bors rollup=never

Nadrieril committed:
Typo

Co-authored-by: varkor <github@varkor.com>

varkor commented:
Thanks! @bors r+

bors commented:
:pushpin: Commit 4c6510bee5ce2b9f9281309dc27f848f2bd3af2c has been approved by `varkor`

<!-- @bors r=varkor 4c6510bee5ce2b9f9281309dc27f848f2bd3af2c -->
<!-- homu: {"type":"Approved","sha":"4c6510bee5ce2b9f9281309dc27f848f2bd3af2c","approver":"varkor"} -->

RalfJung commented on src/librustc_mir_build/hair/pattern/_match.rs:
```diff
@@ -27,8 +29,52 @@
+/// Caveat: this constructors/fields distinction doesn't quite cover every Rust value. For example
+/// a value of type `Rc<u64>` doesn't fit this idea very well, nor do function pointers and various
+/// other things. However, the idea covers everything that can be pattern-matched, and this is all
+/// we need for exhaustiveness checking.
```
Except that function pointers (and raw pointers) can be pattern-matched...
```rust
fn foo() {}

fn bar(x: fn()) -> bool {
    match x {
        foo => true,
        _ => false,
    }
}
```

Nadrieril replied:
Yeahhhh, I wasn't sure whether to include that. I felt like that is the kind of detail I can swipe under the rug in a high-level overview, like uninhabited types. Would you prefer I mention this?

Nadrieril replied:
I guess I could just remove this example actually. Can I still update the PR now?

mark-i-m replied:
@Nadrieril You can update it, but it will need to be reapproved to enter the queue (that's probably ok, because I doubt it would make it to the head of the queue today anyway).

RalfJung replied:
When I wrote the comment, the text explicitly said that function pointers cannot be pattern matched. I am okay with sweeping details under the rug but that was just false information. ;)

Nadrieril replied:
Ah yeah, I understand how you were reading it now. I haven't meant to iloly that they weren't matchable at all

Nadrieril committed:
Don't mention function pointers

See https://github.com/rust-lang/rust/pull/71930#discussion_r426762889

RalfJung commented on src/librustc_mir_build/hair/pattern/_match.rs:
```diff
@@ -64,9 +64,9 @@
-/// other things. However, the idea covers everything that can be pattern-matched, and this is all
-/// we need for exhaustiveness checking.
+/// a value of type `Rc<u64>` doesn't fit this idea very well, nor do various other things.
+/// However, this idea covers everything that can be pattern-matched, and this is all we need for
+/// exhaustiveness checking.
```
```suggestion
/// However, this idea covers every type that can we do exhaustiveness checking for.
```
I think that is accurate -- pointers (function and raw pointers) are never considered exhaustive, right?

Otherwise it should say "covers almost everything" or so, to give a hint that the rabbit hole goes deeper.

Nadrieril committed:
Make caveat more precise

Nadrieril commented:
@varkor Would you mind re-approving, since I edited some comments as per @RalfJung's review?

RalfJung commented:
@bors r=varkor

bors commented:
:pushpin: Commit d7e1d5f0c275c6e876f97a3d95d27a020bc88046 has been approved by `varkor`

<!-- @bors r=varkor d7e1d5f0c275c6e876f97a3d95d27a020bc88046 -->
<!-- homu: {"type":"Approved","sha":"d7e1d5f0c275c6e876f97a3d95d27a020bc88046","approver":"varkor"} -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 9000, this pull request will be tested once the tree is reopened

bors commented:
:hourglass: Testing commit d7e1d5f0c275c6e876f97a3d95d27a020bc88046 with merge 9310e3bd4f425f84fc27878ebf2bda1f30935a63...
<!-- homu: {"type":"BuildStarted","head_sha":"d7e1d5f0c275c6e876f97a3d95d27a020bc88046","merge_sha":"9310e3bd4f425f84fc27878ebf2bda1f30935a63"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=29810)
Approved by: varkor
Pushing 9310e3bd4f425f84fc27878ebf2bda1f30935a63 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"varkor","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=29810"},"merge_sha":"9310e3bd4f425f84fc27878ebf2bda1f30935a63"} -->

eddyb commented on src/librustc_mir_build/hair/pattern/_match.rs:
```diff
@@ -1033,8 +977,263 @@ impl<'tcx> Constructor<'tcx> {
+    /// have not measured if it really made a difference.
+    Slice(&'p [Pat<'tcx>]),
+    Vec(SmallVec<[&'p Pat<'tcx>; 2]>),
+    /// Patterns where some of the fields need to be hidden. `len` caches the number of non-hidden
+    /// fields.
+    Filtered {
+        fields: SmallVec<[FilteredField<'p, 'tcx>; 2]>,
+        len: usize,
```
I'm a bit late to this but I would've called this field `kept_count` instead of `len`.

Nadrieril replied:
I'll change this in my next PR

Nadrieril replied:
Snuck it in here https://github.com/rust-lang/rust/pull/72506
